### PR TITLE
Feat/7/add support of let syntax

### LINKED ISSUE #7: Add support of `let` syntax
let should define variable accessible from 'it'

```ruby
BDDSM.describe "Some" do
  let(:a) { 1 }
  it { expect(a).to eq 1 }
end
```

Also let var should be accessible from another let

```ruby
BDDSM.describe "Some" do
  let(:one) { 1 }
  let(:two) { one + one }
  it { expect(two).to eq 2 }
end
```

### PR DESCRIPTION
**Pull request by lain0**
ref #7 let added

### CODE REVIEW

File: test/let2_spec.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+BDDSM.describe 'Some' do
```
Comment (sergio-fry) [inline test/let2_spec.rb]: Let's combine all let-spec in a single file

File: test/let_spec.rb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+BDDSM.describe 'Some' do
```
Comment (sergio-fry) [inline test/let_spec.rb]: Lets try to use more semantic description

File: lib/bddsm/describe.rb [hunk 3]
```diff
@@ -14,5 +17,11 @@ def run
     def it(&)
       Execution.new(describe: self, &).run
     end
+
+    def let(name, &)
+      self.singleton_class.instance_eval { attr_accessor name.to_sym }
+      @memoized_lets[name] ||= instance_exec(&)
```
Comment (sergio-fry) [inline lib/bddsm/describe.rb]: Looks like we have double execution in here. Please add a test to check if let block execute only once

Comment (lain0) [reply lib/bddsm/describe.rb]: Do you mean  lib/bddsm/describe.rb L23,24
 did [a6f671c] fixed it?

Comment (sergio-fry) [reply lib/bddsm/describe.rb]: Highly likely it has fixed it. lease add a test to check if let block executed only once

File: test/let_spec.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  let(:two) { one + one }
+  it { expect(two).to eq 2 }
+
+  # let variable should be overriden
+  let(:three) { 3 }
+  let(:three) { 4 }
+  it { expect(three).to eq 4 }
+
+  # checks if let block executes only once
+  module Count # rubocop:disable Lint/ConstantDefinitionInBlock
```
Comment (sergio-fry) [inline test/let_spec.rb]: lets disable this cop globally

File: test/let_spec.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
+BDDSM.describe 'BDDSM::Describe.let' do # rubocop:disable Metrics/BlockLength
```
Comment (sergio-fry) [inline test/let_spec.rb]: Let's define some new block length, that we are going to respect

File: test/let_spec.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  module Count # rubocop:disable Lint/ConstantDefinitionInBlock
+    module_function
+
+    @instance_count ||= 0
+
+    def instance_count
+      @instance_count
+    end
+
+    def test(counter)
```
Comment (sergio-fry) [inline test/let_spec.rb]: What this this method is testing? Looks like it increments something

File: test/let_spec.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+      @instance_count += 1
+      counter + 1
+    end
+  end
+
+  # test module class variables works
+  it do
```
Comment (sergio-fry) [inline test/let_spec.rb:38]: This test is unclear. It is impossible to understand, why expect(Count.instance_count).to eq 2

File: test/let_spec.rb [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    Count.test(2)
+    Count.test(3)
+    expect(Count.instance_count).to eq 2
+  end
+
+  let(:count) { Count.test(4) }
```
Comment (sergio-fry) [inline test/let_spec.rb]: Is it possible to use instance instead of static methods?

**Review by sergio-fry — CHANGES_REQUESTED** [review]

File: test/let_spec.rb (hunk 6 above)
Comment (lain0) [reply test/let_spec.rb]: ok i'll rename it

File: test/let_spec.rb (hunk 5 above)
Comment (lain0) [reply test/let_spec.rb]:
```yaml
Metrics/BlockLength:
  Max: 30
  Exclude:
    - test/**/*.rb
  ```
ok?

File: test/let_spec.rb (hunk 7 above)
Comment (lain0) [reply test/let_spec.rb:38]: with instances it seems more clear?

Comment (sergio-fry) [reply test/let_spec.rb:38]: not really

Comment (sergio-fry) [reply test/let_spec.rb:38]: @AlexNimoy @sas2job @sk8higher could you suggest something in here?

File: .rubocop.yml [hunk 9]
```diff
@@ -1,11 +1,26 @@
...
+  Exclude:
+    - test/**/*.rb
+
+Metrics/BlockLength:
+  Max: 30
+  Exclude:
+    - test/**/*.rb
+
+Metrics/MethodLength:
+  Max: 11
```
Comment (sergio-fry) [inline .rubocop.yml]: 10 lines looks enough. Let's try to decompose large methods to keep this cop config default

File: test/let_spec.rb (hunk 7 above)
Comment (sas2job) [reply test/let_spec.rb:38]: have no ideas

Comment (sk8higher) [reply test/let_spec.rb:38]: have no ideas either

File: .rubocop.yml [hunk 10]
```diff
@@ -1,11 +1,26 @@
...
 
-Style/FrozenStringLiteralComment:
-  Enabled: false
+Lint/ConstantDefinitionInBlock:
+  Enabled: true
+  Exclude:
+    - test/**/*.rb
+
+Metrics/BlockLength:
+  Max: 30
```
Comment (sergio-fry) [inline .rubocop.yml]: could we stay with defaults?

File: .rubocop.yml [hunk 11]
```diff
@@ -1,11 +1,26 @@
...
+    - test/**/*.rb
+
+Metrics/MethodLength:
+  Max: 11
 
 Style/Documentation:
   Enabled: false
+
+Style/FrozenStringLiteralComment:
+  Enabled: false
```
Comment (sergio-fry) [inline .rubocop.yml]: let's disallow this frozen literal at all

File: test/let_spec.rb (hunk 7 above)
Comment (sergio-fry) [reply test/let_spec.rb:38]: @lain0 what do you think about something like that?

```ruby
def with_calls_counted
  @calls_count ||= 0
  @calls_count += 1

  yield
end

let(:two) { with_calls_counted { one + 1 } }

it { 
  expect(two).to eq 2
  expect(@calls_count).to eq 1
}
```

**Review by sergio-fry — CHANGES_REQUESTED** [review]

File: test/let_spec.rb (hunk 7 above)
Comment (sergio-fry) [reply test/let_spec.rb:38]: If is is impossible to use instance variable, let's try to use class variables with `@@calls_count`

Comment (lain0) [reply test/let_spec.rb:38]: ok, i used this snippet but inside class  to share class variable between expect instances

**Review by sergio-fry — APPROVED** [review]

**Pull request merged**